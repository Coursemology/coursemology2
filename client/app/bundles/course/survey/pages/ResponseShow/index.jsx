import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { defineMessages, FormattedMessage } from 'react-intl';
import { formatLongDateTime } from 'lib/moment';
import {
  Card,
  CardContent,
  ListSubheader,
  Table,
  TableBody,
  TableCell,
  TableRow,
} from '@mui/material';
import surveyTranslations from 'course/survey/translations';
import { surveyShape, responseShape } from 'course/survey/propTypes';
import { fetchResponse } from 'course/survey/actions/responses';
import LoadingIndicator from 'lib/components/LoadingIndicator';
import ResponseForm, {
  buildInitialValues,
} from 'course/survey/containers/ResponseForm';
import RespondButton from 'course/survey/containers/RespondButton';
import UnsubmitButton from 'course/survey/containers/UnsubmitButton';

const translations = defineMessages({
  notSubmitted: {
    id: 'course.surveys.ResponseShow.notSubmitted',
    defaultMessage: 'Not submitted',
  },
});

const styles = {
  submissionInfoTable: {
    marginTop: 10,
    maxWidth: 600,
  },
};

class ResponseShow extends Component {
  componentDidMount() {
    const {
      dispatch,
      match: {
        params: { responseId },
      },
    } = this.props;
    dispatch(fetchResponse(responseId));
  }

  renderBody() {
    const { survey, response, flags } = this.props;
    if (flags.isLoading) {
      return <LoadingIndicator />;
    }
    const initialValues = buildInitialValues(survey, response);

    return (
      <>
        {this.renderSubmissionInfo()}
        <ListSubheader disableSticky>
          <FormattedMessage {...surveyTranslations.questions} />
        </ListSubheader>
        <ResponseForm readOnly {...{ response, flags, initialValues }} />
      </>
    );
  }

  renderRespondButton() {
    const {
      survey,
      response,
      courseId,
      flags: { canModify, canSubmit, isLoading },
    } = this.props;
    if (!(canModify || canSubmit) || isLoading) {
      return null;
    }

    return (
      <RespondButton
        courseId={courseId}
        surveyId={survey.id}
        responseId={response.id}
        canModify={canModify}
        canSubmit={canSubmit}
        startAt={survey.start_at}
        endAt={survey.end_at}
        submittedAt={response.submitted_at}
      />
    );
  }

  renderSubmissionInfo() {
    const { response } = this.props;
    return (
      <Table style={styles.submissionInfoTable}>
        <TableBody>
          <TableRow>
            <TableCell>Student</TableCell>
            <TableCell>{response.creator_name}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Submitted At</TableCell>
            <TableCell>
              {response.submitted_at ? (
                formatLongDateTime(response.submitted_at)
              ) : (
                <FormattedMessage {...translations.notSubmitted} />
              )}
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Last Updated At</TableCell>
            <TableCell>
              {response.submitted_at ? (
                formatLongDateTime(response.updated_at)
              ) : (
                <FormattedMessage {...translations.notSubmitted} />
              )}
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    );
  }

  renderUnsubmitButton() {
    const {
      response,
      flags: { canUnsubmit, isLoading },
    } = this.props;
    if (!canUnsubmit || isLoading || !response.submitted_at) {
      return null;
    }
    return (
      <span style={{ marginLeft: 12 }}>
        <UnsubmitButton responseId={response.id} />
      </span>
    );
  }

  render() {
    const { survey } = this.props;

    return (
      <>
        {survey.description ? (
          <Card>
            <CardContent
              dangerouslySetInnerHTML={{ __html: survey.description }}
            />
          </Card>
        ) : null}
        {this.renderBody()}
        {this.renderRespondButton()}
        {this.renderUnsubmitButton()}
      </>
    );
  }
}

ResponseShow.propTypes = {
  survey: surveyShape,
  courseId: PropTypes.string.isRequired,
  response: responseShape,
  flags: PropTypes.shape({
    canUnsubmit: PropTypes.bool,
    isLoading: PropTypes.bool.isRequired,
    canModify: PropTypes.bool,
    canSubmit: PropTypes.bool,
  }),
  match: PropTypes.shape({
    params: PropTypes.shape({
      responseId: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  dispatch: PropTypes.func.isRequired,
};

export const UnconnectedResponseShow = ResponseShow;
export default connect((state) => state.responseForm)(ResponseShow);
