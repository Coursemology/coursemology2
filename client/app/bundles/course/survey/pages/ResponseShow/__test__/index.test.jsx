import React from 'react';
import { mount, shallow } from 'enzyme';
import CourseAPI from 'api/course';
import storeCreator from 'course/survey/store';
import ResponseShow, { UnconnectedResponseShow } from '../index';

describe('<ResponseShow />', () => {
  it('allows responses to be saved', async () => {
    const surveyId = '1';
    const responseId = '1';
    const spyFetch = jest.spyOn(CourseAPI.survey.responses, 'fetch');

    mount(
      <ResponseShow
        {...{
          survey: {},
          courseId,
          surveyId,
          match: { params: { responseId } },
        }}
      />,
      buildContextOptions(storeCreator({})),
    );
    await sleep(1);
    expect(spyFetch).toHaveBeenCalled();
  });

  it('shows form and admin buttons if user has permissions and page is loaded', () => {
    const surveyId = 2;
    const responseId = 2;
    const survey = {
      id: surveyId,
      title: 'Survey',
      description: 'Description',
    };
    const responseFormData = {
      response: {
        id: responseId,
        creator_name: 'Staff',
        submitted_at: '2099-12-31T16:00:00.000Z',
        updated_at: '2100-01-12T16:00:00.000Z',
      },
      flags: {
        canModify: true,
        canSubmit: true,
        canUnsubmit: true,
        isResponseCreator: true,
        isLoading: false,
      },
    };
    const urlParams = {
      courseId,
      surveyId: surveyId.toString(),
      match: { params: { responseId: responseId.toString() } },
    };
    const responseShow = shallow(
      <UnconnectedResponseShow
        dispatch={() => {}}
        survey={survey}
        {...responseFormData}
        {...urlParams}
      />,
      buildContextOptions(storeCreator({})),
    );
    expect(responseShow).toMatchSnapshot();
  });

  it('shows only title and loading indicator when loading', () => {
    const surveyId = 2;
    const responseId = 2;
    const responseFormData = {
      response: {
        id: responseId,
        creator_name: 'Student',
      },
      flags: {
        canModify: true,
        canSubmit: true,
        canUnsubmit: true,
        isResponseCreator: true,
        isLoading: true,
      },
    };
    const urlParams = {
      courseId,
      surveyId: surveyId.toString(),
      match: { params: { responseId: responseId.toString() } },
    };
    const responseShow = shallow(
      <UnconnectedResponseShow
        dispatch={() => {}}
        survey={{}}
        {...responseFormData}
        {...urlParams}
      />,
      buildContextOptions(storeCreator({})),
    );
    expect(responseShow).toMatchSnapshot();
  });
});
