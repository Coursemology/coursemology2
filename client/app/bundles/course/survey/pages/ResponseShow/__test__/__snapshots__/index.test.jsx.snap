// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponseShow /> shows form and admin buttons if user has permissions and page is loaded 1`] = `
<Fragment>
  <ForwardRef(Card)>
    <ForwardRef(CardContent)
      dangerouslySetInnerHTML={
        Object {
          "__html": "Description",
        }
      }
    />
  </ForwardRef(Card)>
  <ForwardRef(Table)
    style={
      Object {
        "marginTop": 10,
        "maxWidth": 600,
      }
    }
  >
    <ForwardRef(TableBody)>
      <ForwardRef(TableRow)>
        <ForwardRef(TableCell)>
          Student
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Staff
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
      <ForwardRef(TableRow)>
        <ForwardRef(TableCell)>
          Submitted At
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          01 Jan 2100, 12:00am
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
      <ForwardRef(TableRow)>
        <ForwardRef(TableCell)>
          Last Updated At
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          13 Jan 2100, 12:00am
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
  <ForwardRef(ListSubheader)
    disableSticky={true}
  >
    <FormattedMessage
      defaultMessage="Questions"
      id="course.surveys.questions"
      values={Object {}}
    />
  </ForwardRef(ListSubheader)>
  <ReduxForm
    flags={
      Object {
        "canModify": true,
        "canSubmit": true,
        "canUnsubmit": true,
        "isLoading": false,
        "isResponseCreator": true,
      }
    }
    initialValues={Object {}}
    readOnly={true}
    response={
      Object {
        "creator_name": "Staff",
        "id": 2,
        "submitted_at": "2099-12-31T16:00:00.000Z",
        "updated_at": "2100-01-12T16:00:00.000Z",
      }
    }
  />
  <Connect(RespondButton)
    canModify={true}
    canSubmit={true}
    courseId="1"
    responseId={2}
    submittedAt="2099-12-31T16:00:00.000Z"
    surveyId={2}
  />
  <Connect(UnsubmitButton)
    responseId={2}
  />
</Fragment>
`;

exports[`<ResponseShow /> shows only title and loading indicator when loading 1`] = `
<Fragment>
  <LoadingIndicator />
</Fragment>
`;
